src/pycell/layout.rs

```diff
@@ -0,0 +1,1740 @@
+    #[cfg(Py_3_12)]
+    pub(crate) unsafe fn get_contents_ptr<T: PyClassImpl + PyTypeInfo>(
+        obj: *mut ffi::PyObject,
+        strategy: TypeObjectStrategy<'_>,
+    ) -> *mut PyClassObjectContents<T> {
+        let type_obj = match strategy {
+            TypeObjectStrategy::Lazy(py) => T::type_object_raw(py),
+            TypeObjectStrategy::AssumeInit(_) => {
+                T::try_get_type_object_raw().unwrap_or_else(|| {
+                    panic!(
+                        "type object for {} not initialized",
+                        std::any::type_name::<T>()
+                    )
+                })
+            }
+        };
+        assert!(!type_obj.is_null(), "type object is NULL");
+        debug_assert!(
+            unsafe { ffi::PyType_IsSubtype(ffi::Py_TYPE(obj), type_obj) } == 1,
+            "the object is not an instance of {}",
+            std::any::type_name::<T>()
+        );
+        let pointer = unsafe { ffi::PyObject_GetTypeData(obj, type_obj) };
+        assert!(!pointer.is_null(), "pointer to pyclass data returned NULL");
+        pointer.cast()
+    }
+
+    #[inline(always)]
+    #[cfg(not(Py_3_12))]
+    pub fn panic_unsupported<T: PyClassImpl + PyTypeInfo>() -> ! {
+        assert!(T::OPAQUE);
+        panic!(
+            "The opaque object layout (used by {}) is not supported until python 3.12",
+            std::any::type_name::<T>()
+        );
+    }
+}
+
+/// Utilities for working with [ffi::PyObject] objects that utilise the standard layout for python extensions,
+/// where the base class is placed at the beginning of a `repr(C)` struct.
+#[doc(hidden)]
+pub(crate) mod static_layout {
+    use crate::{
+        impl_::pyclass::{PyClassBaseType, PyClassImpl},
+        type_object::{PyLayout, PySizedLayout},
+    };
+
+    use super::PyClassObjectContents;
+
+    // The layout of a [ffi::PyObject] that uses the static layout
+    #[repr(C)]
+    pub struct PyStaticClassLayout<T: PyClassImpl> {
+        pub(crate) ob_base: <T::BaseType as PyClassBaseType>::StaticLayout,
+        pub(crate) contents: PyClassObjectContents<T>,
+    }
+
+    unsafe impl<T: PyClassImpl> PyLayout<T> for PyStaticClassLayout<T> {}
+
+    /// Layout of a native type `T` with a known size (not opaque)
+    /// Corresponds to [PyObject](https://docs.python.org/3/c-api/structures.html#c.PyObject) from the C API.
+    #[doc(hidden)]
+    #[repr(C)]
+    pub struct PyStaticNativeLayout<T> {
+        ob_base: T,
+    }
+
+    unsafe impl<T, U> PyLayout<T> for PyStaticNativeLayout<U> where U: PySizedLayout<T> {}
+
+    /// a struct for use with opaque native types to indicate that they
+    /// cannot be used as part of a static layout.
+    #[repr(C)]
+    pub struct InvalidStaticLayout;
+
+    /// This is valid insofar as casting a `*mut ffi::PyObject` to `*mut InvalidStaticLayout` is valid
+    /// since `InvalidStaticLayout` has no fields to read.
+    unsafe impl<T> PyLayout<T> for InvalidStaticLayout {}
```

Comment:
Is this just from stopping people from not declaring an opaque layout when inheriting from PyType? It appears to me like it would be better to have another trait for this.